src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il

```diff
@@ -0,0 +1,158 @@
+    01 00 0b 53 65 72 76 69 63 65 61 62 6c 65 04 54
+    72 75 65 00 00
+  ) // "Serviceable", "True"
+  .custom instance void [System.Runtime]System.Reflection.AssemblyCopyrightAttribute::.ctor(string) = ( 01 00 2F C2 A9 20 4D 69 63 72 6F 73 6F 66 74 20   // ../.. Microsoft 
+                                                                                        43 6F 72 70 6F 72 61 74 69 6F 6E 2E 20 20 41 6C   // Corporation.  Al
+                                                                                        6C 20 72 69 67 68 74 73 20 72 65 73 65 72 76 65   // l rights reserve
+                                                                                        64 2E 00 00 )                                     // d...
+  .custom instance void [System.Runtime]System.Reflection.AssemblyCompanyAttribute::.ctor(string) = ( 01 00 15 4D 69 63 72 6F 73 6F 66 74 20 43 6F 72   // ...Microsoft Cor
+                                                                                      70 6F 72 61 74 69 6F 6E 00 00 )                   // poration..
+  .custom instance void [System.Runtime]System.Reflection.AssemblyProductAttribute::.ctor(string) = ( 01 00 1A 4D 69 63 72 6F 73 6F 66 74 C2 AE 20 2E   // ...Microsoft.. .
+                                                                                      4E 45 54 20 46 72 61 6D 65 77 6F 72 6B 00 00 )    // NET Framework..
+  .custom instance void [System.Runtime]System.CLSCompliantAttribute::.ctor(bool) = (
+    01 00 00 00 00
+  ) // false
+  .hash algorithm 0x00008004
+  .ver 4:0:0:0
+}
+.module System.Runtime.CompilerServices.Unsafe.dll
+// MVID: {1E97D84A-565B-49C5-B60A-F31A1A4ACE13}
+.imagebase 0x00400000
+.file alignment 0x00000200
+.stackreserve 0x00100000
+.subsystem 0x0003       // WINDOWS_CUI
+.corflags 0x00000001    //  ILONLY
+// Image base: 0x02ED0000
+
+
+// =============== CLASS MEMBERS DECLARATION ===================
+
+.class public abstract auto ansi sealed beforefieldinit System.Runtime.CompilerServices.Unsafe
+       extends [System.Runtime]System.Object
+{
+  .method public hidebysig static !!T  Read<T>(void* p) cil managed aggressiveinlining
+  {
+        .custom instance void System.Runtime.Versioning.NonVersionableAttribute::.ctor() = ( 01 00 00 00 )
+        .maxstack  8
+        ldarg.0
+        ldobj !!T
+        ret
+  } // end of method Unsafe::Read
+
+  .method public hidebysig static void  Write<T>(void* p,
+                                                 !!T 'value') cil managed aggressiveinlining
+  {
+        .custom instance void System.Runtime.Versioning.NonVersionableAttribute::.ctor() = ( 01 00 00 00 )
+        .maxstack  8
+        ldarg.0
+        ldarg.1
+        stobj !!T
+        ret
+  } // end of method Unsafe::Write
+
+  .method public hidebysig static void  Write<T>(void* p,
+                                                 !!T& 'value') cil managed aggressiveinlining
```

Comment:
Yes, the overload is very much needed. E. g. for compounded value types.
codegen is extremely important to us, so we need both versions. `Unsafe`
use is for advanced users anyway so I don't see the problem.
On Apr 25, 2016 03:57, "Jan Kotas" <notifications@github.com> wrote:

> In
> src/System.Runtime.CompilerServices.Unsafe/src/System.Runtime.CompilerServices.Unsafe.il
> <https://github.com/dotnet/corefx/pull/7966#discussion_r60857131>:
>
> > +        ret
> > +  } // end of method Unsafe::Read
> > +
> > +  .method public hidebysig static void  Write<T>(void* p,
> > +                                                 !!T 'value') cil managed aggressiveinlining
> > +  {
> > +        .custom instance void System.Runtime.Versioning.NonVersionableAttribute::.ctor() = ( 01 00 00 00 )
> > +        .maxstack  8
> > +        ldarg.0
> > +        ldarg.1
> > +        stobj !!T
> > +        ret
> > +  } // end of method Unsafe::Write
> > +
> > +  .method public hidebysig static void  Write<T>(void* p,
> > +                                                 !!T& 'value') cil managed aggressiveinlining
>
> Is this overload really needed? It functionally equivalent to the other
> Write overload. I do not think it should be here.
>
> I understand that one or the other may give better codegen in some cases
> depending on how the JIT is implemented, but having both makes it hard for
> folks to decide which one to use.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/dotnet/corefx/pull/7966/files/0afdc7c9841f3c979ff9be14184c846e56aff380#r60857131>
>